packages/basic-crawler/src/internals/basic-crawler.ts

```diff
@@ -949,14 +973,28 @@ export class BasicCrawler<
             await source.markRequestHandled(request);
             this.stats.failJob(request.id || request.uniqueKey);
 
-            // @ts-expect-error It is assignable, but TS says otherwise...
-            const castedErrorContext = crawlingContext as ErrorContext;
-            castedErrorContext.error = error;
-
             await this._handleFailedRequestHandler(castedErrorContext); // This function prints an error message.
         }
     }
 
+    protected async _handlePreReclaimRequestHandler(crawlingContext: ErrorContext, source: RequestList | RequestQueue): Promise<void> {
+        if (this.preReclaimRequestHandler) {
+            await this.preReclaimRequestHandler(crawlingContext);
+        } else {
+            const { id, url, retryCount } = crawlingContext.request;
+            // We don't want to see the stack trace in the logs by default, when we are going to retry the request.
+            // Thus, we print the full stack trace only when CRAWLEE_VERBOSE_LOG environment variable is set to true.
+            this.log.warning(
+                `Reclaiming failed request back to the list or queue. ${!process.env.CRAWLEE_VERBOSE_LOG
+                    ? crawlingContext.error
+                    : crawlingContext.error.stack
+                }`,
+                { id, url, retryCount },
+            );
+            await source.reclaimRequest(crawlingContext.request);
```

Comment:
Why is the reclaim inside the `else` clause? The handler should just give the user an opportunity do do something and then continue the same way as before.

If the user wants to prevent the request from being reclaimed, they should set `request.noRetry = true` and the retry mechanism of basic crawler should mark the request as handled.

Reply:
The use-case for the handler is generic, really: "Let the user do whatever with the request, if there's an error". Because currently, there's no way to do this. If there's an error in the process, the request will immediately go into internal code.

So reclaiming or not reclaiming is not the concern. Quite the contrary, because we don't want to complicate things for the user by forcing them to reclaim manually.

```js
errorHandler: () => { console.log('there was an error') }
``` 

is a valid handler.

Reply:
Yup, get it now 👍